[shunjieee] iP

shunjieee opened the pull request:
# Xavier, your ~~Financial~~ Task Advisor 

> Good Morning. Hustler! ~ Xavier

Xavier frees your mind ~~**(I mean take your precious time)**~~ of having remember things you need to do. It's,

* text-based
* easy to learn
* ~~fast~~ _SUPER DUPER_ FAST

All you need to do is,

1. download it from [here](https://github.com/shunjieee/ip/releases/tag/A-Jar).
2. double-click it.
3. add you tasks.
4. let it manage your tasks for you 😇.

And it is **FREE!!!**

Features:
- [x] Manage tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon... or never 😅)

---
If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

Commit:
Duke.java: Add Level-0 requirements

Commit:
Duke.java: Add Level-1 requirements

The chatbot now take user inputs.

Implementing echo functionality allows user to interact with the
chatbot.

Commit:
Duke.java: Add Level-2 requirements

User inputs are tasks to be added into the list. Inputting
"list" allows return of the tasks listed in chronological order.

Let's use a data structure (i.e. ArrayList) to store the tasks.

ArrayList allows random access of tasks in the array.

Commit:
Add helperpackage and Task.java

Commit:
Duke.java: Add Level-3 requirements

Commit:
Add Deadline.java

Commit:
Add Event.java

Commit:
Add ToDo.java

Commit:
ToDo.java

Commit:
Update test cases and expected results

Commit:
Deadline.java: Add DukeException for invalid inputs

Commit:
Event.java: Add DukeException for invalid inputs

Commit:
Add DukeException.java

Commit:
Duke.java: Add try-catch for exceptions

Commit:
Duke.java: Add Level-6 requirements

Commit:
Merge commit '02c1ec31378fffb8932bc84b33ea9c024e97ce70'

* commit '02c1ec31378fffb8932bc84b33ea9c024e97ce70':
  Add ToDo.java

JiaXinEu commented on src/main/java/Duke.java:
```diff
@@ -1,3 +1,14 @@
+import helperpackage.Deadline;
+import helperpackage.DukeException;
+import helperpackage.Event;
+import helperpackage.Task;
+import helperpackage.ToDo;
+
+import java.util.LinkedList;
+import java.util.NoSuchElementException;
+import java.util.Scanner;
+import java.util.StringTokenizer;
```
```suggestion
import java.util.LinkedList;
import java.util.NoSuchElementException;
import java.util.Scanner;
import java.util.StringTokenizer;

import helperpackage.Deadline;
import helperpackage.DukeException;
import helperpackage.Event;
import helperpackage.Task;
import helperpackage.ToDo;
```

JiaXinEu commented on src/main/java/helperpackage/Task.java:
```diff
@@ -0,0 +1,50 @@
+    /** Task description, also the user input */
+    protected String description;
+    /** Status of the task, done or not done */
+    protected boolean isDone;
```
I like how you name the Boolean variable 😃

JiaXinEu commented on src/main/java/Duke.java:
```diff
@@ -6,5 +17,150 @@ public static void main(String[] args) {
+    public static void delete (LinkedList<Task> taskList, String cmd, StringTokenizer st) throws IndexOutOfBoundsException,
+            NumberFormatException, NoSuchElementException {
+
+    int index = Integer.parseInt(st.nextToken());
+    Task t = taskList.remove(index - 1);
+
+    System.out.println("Noted, I've removed this task:");
+    System.out.println(" " + t.toString());
+    System.out.println("Now you have " + taskList.size() + " tasks in the list.");
+}
```
```suggestion
        int index = Integer.parseInt(st.nextToken());
        Task t = taskList.remove(index - 1);
        
        System.out.println("Noted, I've removed this task:");
        System.out.println(" " + t.toString());
        System.out.println("Now you have " + taskList.size() + " tasks in the list.");
    }
```

JiaXinEu commented on src/main/java/Duke.java:
```diff
@@ -6,5 +17,150 @@ public static void main(String[] args) {
     }
+
+    // Level-3: mark & unmark
+    public static void changeStatus(LinkedList<Task> taskList, String cmd, StringTokenizer st) throws IndexOutOfBoundsException,
```
I like how you use line wrapping at appropriate place of this line

JiaXinEu reviewed (commented):
LGTM 👍 
Overall, I find your code easy to read and most part follows the coding standards.

ljy0422 commented on src/main/java/Duke.java:
```diff
@@ -6,5 +17,150 @@ public static void main(String[] args) {
+    // Level-4: ToDo, Deadline, Event
+    public static void addTask(LinkedList<Task> taskList, String cmd, String userInput) throws DukeException {
+        int firstSpaceIndex = userInput.indexOf(" ");
+        String description = userInput.substring(firstSpaceIndex + 1);
```
Here the `description` means the whole part following the command type, while in the Task class another `description` means solely the description part of a task, which would be a little misleading or ambiguous. Would it be better to change the variable name here?

ljy0422 commented on src/main/java/Duke.java:
```diff
@@ -6,5 +17,150 @@ public static void main(String[] args) {
+    public static void delete (LinkedList<Task> taskList, String cmd, StringTokenizer st) throws IndexOutOfBoundsException,
+            NumberFormatException, NoSuchElementException {
+
+    int index = Integer.parseInt(st.nextToken());
```
The name `index` doesn't explain its meaning accurately. Could you try to specify what the index stands for, like the `firstSpaceIndex` in your code above?

ljy0422 reviewed (commented):
After reviewing your code for naming standards, no significant issues are identified. Nonetheless, there are a couple of spots where tweaking the names could make things a bit clearer. Keep up the great work! :)))

Commit:
Add toData() function

This function allows a different formatting that is used in the
text file, e.g. E | 1 | concert | Aug 2nd 4-6pm, as compared to
the toString method.

This format is also in accordance with how the text file will
be read when the programme runs.

Let's create a new function to give an output with a specific format

Commit:
Add branch-Level-7 requirements

Data can be read from a text file when the program is
initialised. The text file is updated when the program
terminated.

Let's
* create an instance of File to locate the text file
* handle errors when locating the file, i.e. file / path
does not exist
* read existing data from the file and add them into the
taskList, if any
* update the text file when the program terminates

Commit:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Add branch-Level-7 requirements
  Add toData() function

Commit:
Change deadline from String to LocalDateTime

More functionalities can be used by storing the deadline
as LocalDateTime as compared to String

Let's
* Create one formatter for user inputs
* Create one formatter for output and reading from text file
* Parse String into LocalDateTime
* Handle error if the input does not match the formats

Commit:
Change startAndEndTime from String to LocalDateTime

More functionalities can be used by storing the start time and end time
as LocalDateTime as compared to String

Let's
* Create one formatter for user inputs
* Create one formatter for output and reading from text file
* Parse String into LocalDateTime
* Handle error if the input does not match the formats

Commit:
Add branch-Level-8 requirements

After changing Deadline.java and Event.java to use LocalDateTime
instead of String to store date and time, Duke.java has to handle
exceptions thrown by those two classes.

Let's add try-catch block to handle exceptions thrown when creating
an instance of Deadline or Event.

Commit:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Add branch-Level-8 requirements
  Change startAndEndTime from String to LocalDateTime
  Change deadline from String to LocalDateTime

Commit:
Update format of input

The new format input allow easier parsing.

Let's use StringTokeniser to read and process the user inputs.

Using StringTokeniser allows a neater string processing as compared to slicing.

Commit:
Merge branch 'master' into branch-A-MoreOOP

* master:
  Update format of input

Commit:
Add Ui.java and Storage.java

Additional class and functions have been created for abstraction.

Let's:
* Put most user-interaction functions in Ui.java
* Put load and store data functions in Storage.java
* Check for file and directory existence in Storage.java

Using abstraction allows simplification and encapsulation of complex
functions.

Commit:
Add TaskList.java

Additional class and functions have been created for abstraction for the management of data.

Let's abstract functions with regards to data manipulation in Storage.java.

Using abstraction allows simplification and encapsulation of complex functions.

Commit:
Add Parser.java

Additional class and functions aid in parsing of user input.

The parse command returns a command of a specific type according to the user input.

Let's abstract the if-else statements and simplify them in Parser.java.

Using abstraction allows simplification and encapsulation of parsing and returning of the different commands.

Commit:
Add command package

Additional classes and functions have been created for abstraction of individual command functions.

Let's
* create command package to store Command.java and its subclasses
* create Command.java and abstract functions for subclasses
* implement subclasses execute function logic
* add isExit function to differentiate ExitCommand.java with other subclasses

Using abstraction allows simplification and encapsulation of individual command functions.

Commit:
Merge branch 'branch-A-MoreOOP'

* branch-A-MoreOOP:
  Add command package
  Add Parser.java
  Add TaskList.java
  Add Ui.java and Storage.java

Commit:
Merge branch 'add-gradle-support' into branch-Level-9

Commit:
Revert "Merge branch 'add-gradle-support' into branch-Level-9"

This reverts commit 273756d9994a5b98859fad0a6dd6cab388e88a78, reversing
changes made to ed6a9e196a0cdac600f9147225fcf639de225549.

Commit:
Add Level-9 requirements

Commit:
Update files to match coding standards

Commit:
Update files to include JavaDocs and comments

Commit:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Update files to include JavaDocs and comments

Commit:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Update files to match coding standards

# Conflicts:
#	src/main/java/common/DukeException.java
#	src/main/java/common/Parser.java
#	src/main/java/common/Storage.java

Commit:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add Level-9 requirements
  Revert "Merge branch 'add-gradle-support' into branch-Level-9"
  Bump gradle and lib version
  Add Gradle support
  docs/README.md: Tweak document template

# Conflicts:
#	src/main/java/Duke.java
#	src/main/java/common/Parser.java

Commit:
Storage.java: Resolve merge conflict

Commit:
Update gradle filepath

Commit:
Merge branch 'add-gradle-support'

* add-gradle-support:
  Update gradle filepath

# Conflicts:
#	build.gradle

Commit:
Add JUnit test for Deadline.java and Event.java

Commit:
Update coding standard according to checkstyle

minghowC reviewed (commented):
Overall, I found your code is easy to read. LGTM, just a few nits to fix.
For simplicity sake, perpahs line 72 "`boolean isDone = Integer.parseInt(tasks[1]) == 1 ? true : false;` can be simplified `to boolean isDone = Integer.parseInt(tasks[1]) == 1;`? Also, perhaps a more intuitive constant variable name here for "T, "D", and "E"? 
Pardon my PR at this point of time.

Commit:
Add GUI and Level-10 requirement

Commit:
Add GUI and Level 10 requirement

Commit:
Revert "Add GUI and Level 10 requirement"

This reverts commit e7e3ceed9c5705285e7adea6d3ddb131fd165564.

Commit:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Revert "Add GUI and Level 10 requirement"
  Add GUI and Level 10 requirement

Commit:
Add assertions for add, delete, mark and unmark

Assertions can allow programmers to detect possible bugs in the code.

Let's add assertions after the main logic of add, delete, mark and unmark commands.

Using assertion is preferable over exceptions in this situation as the errors are not due to invalid user input or the environment.

shunjieee committed:
Merge pull request #2 from shunjieee/branch-A-Assertions

Add assertions for add, delete, mark and unmark

Commit:
Improve code quality

Minor adjustment made to improve code readability and quality, adhering to principles such as SLAP and KISS.

Let's,
* abstract duplicate codes
* run checkstyleMain for any violations of conventions
* Apply SLAP and KISS principle

shunjieee committed:
Merge pull request #3 from shunjieee/branch-A-CodeQuality

Improve code quality

Commit:
Merge branch 'master' of github.com:shunjieee/ip

* 'master' of github.com:shunjieee/ip:
  Improve code quality

Commit:
Add sort command

Allow user to sort the list base on the task's completion status.

Let's
* add SortCommand.java
* update Parser.java to include "sort"

Commit:
Update GUI and add Ui.png

Commit:
README.md: Add User Guide

shunjieee committed:
Update README.md

shunjieee committed:
Update README.md

shunjieee committed:
Update README.md

shunjieee committed:
Update README.md

shunjieee committed:
Update README.md

Commit:
Merge branch 'master' of github.com:shunjieee/ip

* 'master' of github.com:shunjieee/ip:
  Update README.md
  Update README.md
  Update README.md
  Update README.md
  Update README.md

shunjieee committed:
Update README.md
